bundles/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/internal/handler/PHCBridgeHandler.java

```diff
@@ -104,209 +112,358 @@ public void initialize() {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "Serial port '" + port + "' could not be found.");
+                    "Could not configure serial port " + port + ": " + e.getMessage());
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Failed to get input or output stream for serialPort: " + e.getMessage());
+            logger.debug("Failed to get inputstream for serialPort", e);
         }
     }
 
     /**
-     * Puts the available data on serial port into a buffer and transfer it to the processing method.
+     * Reads the data on serial port and puts it into the internal buffer.
      */
     @Override
     public void serialEvent(SerialPortEvent event) {
-        if (event.getEventType() != SerialPortEvent.DATA_AVAILABLE) {
-            return;
+        if (event.getEventType() == SerialPortEvent.DATA_AVAILABLE && serialIn != null) {
+            try {
+                byte[] bytes = new byte[serialIn.available()];
+                serialIn.read(bytes);
+
+                buffer.offer(bytes);
+
+                if (logger.isTraceEnabled()) {
+                    logger.trace("buffer offered {}", HexUtils.bytesToHex(bytes, " "));
+                }
+            } catch (IOException e) {
+                logger.warn("Error on reading input stream to internal buffer", e);
+            }
         }
+    }
 
-        byte[] buffer;
+    /**
+     * process internal incoming buffer (recognize on read messages)
+     */
+    private void processReceivedBytes() {
+        int faultCounter = 0;
 
         try {
-            if (serialIn.available() <= 0) {
-                return;
-            }
+            byte module = buffer.get();
+
+            while (true) {
+                // Recognition of messages from byte buffer.
+                // not a known module address
+                if (!modules.contains(module)) {
+                    module = buffer.get();
+                    continue;
+                }
 
-            Thread.sleep(5);
-            buffer = new byte[serialIn.available()];
-            serialIn.read(buffer);
+                if (logger.isDebugEnabled()) {
+                    logger.debug("get module: {}", new String(HexUtils.byteToHex(module)));
+                }
 
-            if (messageFragment != null) {
-                processInputStream(ArrayUtils.addAll(messageFragment, buffer));
-            } else {
-                processInputStream(buffer);
+                byte sizeToggle = buffer.get();
+
+                // read length of command and check if makes sense
+                if ((sizeToggle < 1 || sizeToggle > 3) && ((sizeToggle & 0xFF) < 0x81 || (sizeToggle & 0xFF) > 0x83)) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("get invalid sizeToggle: {}", new String(HexUtils.byteToHex(sizeToggle)));
```

Comment:
Is this new String call needed? I would be surprised?

Reply:
I tried it again, but without new String the byte array is written out as array and new String creates an String from the characters in the array. And this is what I want to log.

Reply:
This is indeed a newer method which is not suitable for logging, it was added later on in to the HexUtils